src/native/corehost/fxr/framework_info.cpp

```diff
@@ -43,49 +43,58 @@ bool compare_by_name_and_version(const framework_info &a, const framework_info &
 
     int32_t hive_depth = 0;
 
-    for (pal::string_t dir : hive_dir)
+    for (const pal::string_t& dir : hive_dir)
     {
         auto fx_shared_dir = dir;
         append_path(&fx_shared_dir, _X("shared"));
 
-        if (pal::directory_exists(fx_shared_dir))
+        if (!pal::directory_exists(fx_shared_dir))
+            continue;
+
+        std::vector<pal::string_t> fx_names;
+        if (fx_name != nullptr)
         {
-            std::vector<pal::string_t> fx_names;
-            if (fx_name.length())
-            {
-                // Use the provided framework name
-                fx_names.push_back(fx_name);
-            }
-            else
-            {
-                // Read all frameworks, including "Microsoft.NETCore.App"
-                pal::readdir_onlydirectories(fx_shared_dir, &fx_names);
-            }
+            // Use the provided framework name
+            fx_names.push_back(fx_name);
+        }
+        else
+        {
+            // Read all frameworks, including "Microsoft.NETCore.App"
+            pal::readdir_onlydirectories(fx_shared_dir, &fx_names);
+        }
 
-            for (pal::string_t fx_name_local : fx_names)
-            {
-                auto fx_dir = fx_shared_dir;
-                append_path(&fx_dir, fx_name_local.c_str());
+        for (const pal::string_t& fx_name_local : fx_names)
+        {
+            auto fx_dir = fx_shared_dir;
+            append_path(&fx_dir, fx_name_local.c_str());
+
+            if (!pal::directory_exists(fx_dir))
+                continue;
+
+            trace::verbose(_X("Gathering FX locations in [%s]"), fx_dir.c_str());
 
-                if (pal::directory_exists(fx_dir))
+            std::vector<pal::string_t> versions;
+            pal::readdir_onlydirectories(fx_dir, &versions);
+            for (const pal::string_t& ver : versions)
+            {
+                // Make sure we filter out any non-version folders.
+                fx_ver_t parsed;
+                if (!fx_ver_t::parse(ver, &parsed, false))
+                    continue;
+
+                // Check that the framework's .deps.json exists.
+                pal::string_t fx_version_dir = fx_dir;
+                append_path(&fx_version_dir, ver.c_str());
+                if (!library_exists_in_dir(fx_version_dir, fx_name_local + _X(".deps.json"), nullptr))
```

Comment:
Nit: Maybe as a future cleanup - rename the helper as it's not about libraries... just files.

Reply:
Will plan to do that in a future cleanup.